packages/hardhat-core/src/internal/hardhat-network/stack-traces/vm-debug-tracer.ts

```diff
@@ -0,0 +1,582 @@
+        const addressStorage = this._addressToStorage[address] ?? {};
+        structLog.storage = {
+          ...addressStorage,
+          ...structLog.storage,
+        };
+        this._addressToStorage[address] = {
+          ...structLog.storage,
+        };
+
+        // sometimes the memSize has the correct value
+        // for the memory length, in those cases we increase
+        // the memory to reflect this
+        if (structLog.memSize > structLog.memory.length) {
+          const wordsToAdd = structLog.memSize - structLog.memory.length;
+          for (let k = 0; k < wordsToAdd; k++) {
+            structLog.memory.push(EMPTY_MEMORY_WORD);
+          }
+        }
+
+        if (i === 0) {
+          continue;
+        }
+
+        let previousStructLog = nestedStructLogs[nestedStructLogs.length - 2];
+
+        if (Array.isArray(previousStructLog)) {
+          previousStructLog = nestedStructLogs[nestedStructLogs.length - 3];
+        } else {
+          // if the previous log is not a message, we update its gasCost
+          // using the gas difference between both steps
+          previousStructLog.gasCost = previousStructLog.gas - structLog.gas;
+        }
+
+        assertHardhatInvariant(
+          !Array.isArray(previousStructLog),
+          "There shouldn't be two messages one after another"
+        );
+
+        // the increase in memory size of a revert is immediately
+        // reflected, so we don't treat it as a memory expansion
+        // of the previous step
+        if (structLog.op !== "REVERT") {
+          const memoryLengthDifference =
+            structLog.memory.length - previousStructLog.memory.length;
+          for (let k = 0; k < memoryLengthDifference; k++) {
+            previousStructLog.memory.push(EMPTY_MEMORY_WORD);
+          }
+        }
+      } else {
+        const subMessage: DebugMessage = messageOrStructLog;
+
+        const lastStructLog = nestedStructLogs[nestedStructLogs.length - 1];
+
+        assertHardhatInvariant(
+          !Array.isArray(lastStructLog),
+          "There shouldn't be two messages one after another"
+        );
+
+        const isDelegateCall = lastStructLog.op === "DELEGATECALL";
+
+        const messageNestedStructLogs = await this._messageToNestedStructLogs(
+          subMessage,
+          isDelegateCall ? address : subMessage.to
+        );
+
+        nestedStructLogs.push(messageNestedStructLogs);
+      }
+    }
+
+    return nestedStructLogs;
+  }
+
+  private _getMemory(step: InterpreterStep): string[] {
+    const memory = Buffer.from(step.memory)
+      .toString("hex")
+      .match(/.{1,64}/g);
+
+    const result = memory === null ? [] : [...memory];
+
+    return result;
+  }
+
+  private _getStack(step: InterpreterStep): string[] {
+    const stack = step.stack
+      .slice()
+      .map((el: BN) => el.toString("hex").padStart(64, "0"));
+    return stack;
+  }
+
+  private async _stepToStructLog(step: InterpreterStep): Promise<StructLog> {
```

Comment:
Is this logic affected by Berlin?

Reply:
Uhm, I think `_callConstantGas` changes, yeah. Damn. Does the VM exposes some way of knowing if some address was accessed or not?

Reply:
There is an `isWarmedAddress` method in the state manager. I'll use that.

Reply:
Fixed in 2f80919914f6afbf2b34527526fd296be29abb80